doc/api/process.markdown

```diff
@@ -179,6 +179,110 @@ this, you can either attach a dummy `.catch(() => { })` handler to
+Node.js can emit warnings whenever it detects: (a) the use of deprecated 
+APIs, (b) known security vulnerabilities that are the result of poor or 
+outdated programming practice, and (c) bad coding practices that could lead to 
+sub-optimal application performance or bugs.
+
+The event handler for `'warning'` events is called with a single `warning` 
+argument whose value is an `Error` object. There are three key properties that 
+describe the warning:
+
+* `name` - The name of the warning (currently one of `DeprecationWarning`,
+  `SecurityWarning`, or `BadPracticeWarning`).
+* `message` - A system-provided description of the warning.
+* `stack` - A stack trace to the location in the code where the warning was
+  issued.
+
+```js
+process.on('warning', (warning) => {
+  console.log(warning.name);    // Print the warning name
+  console.log(warning.message); // Print the warning message
+  console.log(warning.stack);   // Print the stack trace
+});
+```
+
+By default, Node.js will print Process Warnings to `stderr`. The `--no-warnings`
+command-line option can be used to suppress the default console output but the
+`'warning'` event will still be emitted by the `process` object.
+
+The following example illustrates the default deprecation warning that is
+printed to `stderr` when a deprecated API is used:
+
+```
+$ node
+> util.debug('foo');
+DEBUG: foo
+undefined
+> DeprecationWarning: (node:94742) util.debug is deprecated. Use
+console.error instead.
+```
+
+In contrast, the following example turns off the default warning output and
+adds a custom handler to the `'warning'` event:
+
+```
+$ node --no-warnings
+> var p = process.on('warning', (warning) => console.log('DEPRECATED API!'));
+> util.debug('foo');
+DEBUG: foo
+undefined
+> DEPRECATED API!
+```
+
+The `--trace-warnings` command-line option can be used to have the default
+console output for warnings include the full stack trace of the warning.
+
+### Warning Types
+
+Node.js can produce three types of warnings:
+
+* `SecurityWarning` - a warning indicating that sub-optimal or unsafe security
+  parameters are being used. For instance, when a weak algorithm is selected or
+  when weak keys are used.
+* `BadPracticeWarning` - a warning indicating that parts of the Node.js API are
+  being used in a manner that is not recommended or has known issues.
+* `DeprecationWarning` - a warning indicating that a deprecated feature is 
+  being used.
+  
+### Emitting Custom Warnings
+
+The [`process.emitWarning(warning[, warned])`][process_emit_warning] method can 
+be used issue custom or application specific warnings.
+
+```js
+// Emit a warning using a string...
+process.emitWarning('Warning! Something happened!');
+
+// Emit a warning using an object...
+process.emitWarning({
+  name: 'CustomWarning', 
+  message: 'Warning! Something Happened!'
+})
+
+// Emit a warning using an Error object...
+const util = require('util');
+
+function CustomWarning(message) {
```

Comment:
They would, yes, but are we at a point yet where we're recommending using
ES6 classes? If so, I have no problem with that change.
On Feb 27, 2016 8:48 PM, "Chris Dickinson" <notifications@github.com> wrote:

> In doc/api/process.markdown
> <https://github.com/nodejs/node/pull/4782#discussion_r54344159>:
>
> > +be used issue custom or application specific warnings.
> > +
> > +```js
> > +// Emit a warning using a string...
> > +process.emitWarning('Warning! Something happened!');
> > +
> > +// Emit a warning using an object...
> > +process.emitWarning({
> > +  name: 'CustomWarning',
> > +  message: 'Warning! Something Happened!'
> > +})
> > +
> > +// Emit a warning using an Error object...
> > +const util = require('util');
> > +
> > +function CustomWarning(message) {
>
> It seems like using ES6 classes might simplify this example a bit:
>
> class CustomWarning extends Error {
>   constructor (msg) {
>     super(`${prefix}${msg}`);
>     Error.captureStackTrace(this, CustomWarning);
>   }
> }
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nodejs/node/pull/4782/files#r54344159>.
>